### PR TITLE
DocAI incubator tools

### PR DESCRIPTION
**Pull request by dmoonat**
Tools which will be helpful for the Doc AI users.

### CODE REVIEW

File: DocAI Incubator Tools/best-practices/Child_Entity_Tag_Using_Header_Keyword.ipynb [hunk 1]
```diff
@@ -0,0 +1,726 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Child_Entity_Tag_Using_Header_Keyword.ipynb:1]: <p>Maybe rephrase this since they don't need to copy/paste in a notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Child_Entity_Tag_Using_Header_Keyword.ipynb:1]: <p>Can you rephrase this? I'm not sure I understand</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Child_Entity_Tag_Using_Header_Keyword.ipynb:1]: <p>Most of the functions in this code block are incredibly long and difficult to read. Could they be split up into multiple functions to simplify readability?</p><p> </p><p>When working with the Document object, it's best to use the object type rather than dicts.</p><p> </p><p>There's also no type annotations for the functions - which should be added.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Child_Entity_Tag_Using_Header_Keyword.ipynb:1]: <p>Remove Go link</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

File: DocAI Incubator Tools/best-practices/Document_AI_Parser_Result_Merger.ipynb [hunk 2]
```diff
@@ -0,0 +1,686 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Document_AI_Parser_Result_Merger.ipynb]: <p>Change the parameter names to be displayed in code syntax</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Document_AI_Parser_Result_Merger.ipynb]: <p>Similar comment to above, this needs to be refactored for readability, use the document object instead of dicts, and add typing.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

File: DocAI Incubator Tools/best-practices/HITL_Rejected_Documents_Tracking.ipynb [hunk 3]
```diff
@@ -0,0 +1,415 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/HITL_Rejected_Documents_Tracking.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #12.    from google.cloud import documentai_v1beta3 as documentai
</pre><p>Is there a reason beta3 is used instead of v1?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/HITL_Rejected_Documents_Tracking.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #90.            # opts = {"api_endpoint": "us-autopush-documentai.sandbox.googleapis.com"}
</pre><p>Remove commented out code - Same throughout</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/HITL_Rejected_Documents_Tracking.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #109.            hitl_op_id = hitl_path.split("/")[-1]
</pre><p>You don't need to do this, just send hitl_path to the get_operation request</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: DocAI Incubator Tools/best-practices/Key_Value_Pair_Entity_Conversion.ipynb [hunk 4]
```diff
@@ -0,0 +1,459 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Key_Value_Pair_Entity_Conversion.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #36.    def entity_data(formField_data: dict, page_number: int):
</pre><p>This whole function should use the entity object instead of a dict</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Key_Value_Pair_Entity_Conversion.ipynb]: <p>Remove go link</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

File: DocAI Incubator Tools/best-practices/Pre_Post_HITL_Bounding_Box_Mismatch.ipynb [hunk 5]
```diff
@@ -0,0 +1,813 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Pre_Post_HITL_Bounding_Box_Mismatch.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #65.    def file_names(file_path):
</pre><p>Quite a few functions throughout these notebooks are reused, they could be moved into a utilities.py file to simplify the notebooks</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: DocAI Incubator Tools/best-practices/Pre_and_Post_HITL_Visualization.ipynb [hunk 6]
```diff
@@ -0,0 +1,962 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Pre_and_Post_HITL_Visualization.ipynb]: <pre class="ql-syntax" spellcheck="false">Line #154.    def bbox_maker(boundingPoly):
</pre><p>use snake_case for variable names</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

File: DocAI Incubator Tools/best-practices/Removing_Empty_Bounding_Boxes.ipynb [hunk 7]
```diff
@@ -0,0 +1,420 @@
+{
```
Comment (holtskinner) [inline DocAI Incubator Tools/best-practices/Removing_Empty_Bounding_Boxes.ipynb]: <p>Similar comment to before, a lot of this code is very difficult to read, could be refactored for easier understanding</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/document-ai-samples/pull/640/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

**Comment by holtskinner** [comment]
General comment - Quite a lot of these code samples are difficult to understand due to high levels of nesting and super long functions. I'd recommend refactoring most of them to make it easier for users to read and understand what is happening.

Also, @dmoonat I'm not sure how many of these are from this PR direct from the incubator team, https://github.com/GoogleCloudPlatform/document-ai-samples/pull/624 but they are in the process of trying to add their samples to the repo. Just not sure how prioritized this is, and most of their samples are currently in PDF format instead of notebooks.

I would just rewrite them myself, but there's a lot to parse through, and I do not have time to work on them.

**Comment by holtskinner** [comment]
Thank you @dmoonat for the updates/improvements!

I'm going to make a minor change to the folder structure to remove spaces in the names, then I'll approve and merge this.

**Review by holtskinner — APPROVED** [review]

File: incubator-tools/best-practices/utilities/utilities.py
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:1]: General comment - Add type annotations for all parameters and return types

File: incubator-tools/best-practices/utilities/utilities.py [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+
+    Returns:
+        documentai.Document: A DocumentAI Document proto representation of the downloaded JSON.
+    """
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(blob_name_with_prefix_path)
+
+    contents = blob.download_as_string()
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:182]: I think this should be `blob.download_as_bytes()` then dont use the `.decode()` method.

File: incubator-tools/best-practices/utilities/utilities.py [hunk 9]
```diff
@@ -0,0 +1,743 @@
...
+                            round(max(y1), 8),
+                        ]
+                        df.loc[len(df.index)] = [
+                            subentity.type_,
+                            subentity.mention_text,
+                            bbox,
+                            page,
+                        ]
+                    except Exception:
+                        continue
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:258]: This is all duplicated below. Can it be moved into a separate function?

File: incubator-tools/best-practices/utilities/utilities.py [hunk 10]
```diff
@@ -0,0 +1,743 @@
...
+        ):
+            if (
+                df_compare.iloc[i]["Pre_HITL_Output"]
+                == df_compare.iloc[i]["Post_HITL_Output"]
+            ):
+                match_string = "TP"
+            else:
+                match_string = "FP"
+        else:
+            match_string = "Something went Wrong."
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:613]:
```suggestion
        match_string = ""
        not_found_string = "Entity not found."
        pre_output = df_compare.iloc[i]["Pre_HITL_Output"]
        post_output = df_compare.iloc[i]["Post_HITL_Output"]
        
        if pre_output == not_found_string and post_output == not_found_string:
            match_string = "TN"
        elif pre_output != not_found_string and post_output == not_found_string:
            match_string = "FN"
        elif pre_output == not_found_string and post_output != not_found_string:
            match_string = "FP"
        elif pre_output != not_found_string and post_output != not_found_string:
            match_string = "TP" if pre_output == post_output else "FP"
        else:
            match_string = "Something went Wrong."
```

File: incubator-tools/best-practices/utilities/utilities.py [hunk 11]
```diff
@@ -0,0 +1,743 @@
...
+
+    Example:
+        schema = get_document_schema("eu", "123456", "processor123", "version123")
+    """
+    # Choose the endpoint based on the provided location.
+    opts = {}
+    if location == "eu":
+        opts = {"api_endpoint": "eu-documentai.googleapis.com"}
+    else:
+        opts = {"api_endpoint": "us-documentai.googleapis.com"}
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:650]:
```suggestion
    # You must set the `api_endpoint` if you use a location other than "us".
    opts = ClientOptions(api_endpoint=f"{location}-documentai.googleapis.com")
```

Note: You'll also need to add this at the top of the file:

```py
from google.api_core.client_options import ClientOptions
```

File: incubator-tools/best-practices/utilities/utilities.py [hunk 12]
```diff
@@ -0,0 +1,743 @@
...
+    # Initialize the DocumentAI client.
+    client = documentai.DocumentProcessorServiceClient(client_options=opts)
+
+    # Construct the request.
+    name = f"projects/{project_number}/locations/{location}/processors/\
+    {processor_id}/processorVersions/{processor_version_id}"
+    request = documentai.GetProcessorVersionRequest(name=name)
+
+    # Fetch the processor version details.
+    response = client.get_processor_version(request=request)
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:661]:
```suggestion
    # The full resource name of the processor version
    # e.g.: projects/project_id/locations/location/processors/processor_id/processorVersions/processor_version_id
    name = client.processor_version_path(
        project_id, location, processor_id, processor_version_id
    )
    # Fetch the processor version details.
    response = client.get_processor_version(name=name)
```

File: incubator-tools/best-practices/utilities/utilities.py [hunk 13]
```diff
@@ -0,0 +1,743 @@
...
+            images[0].save(
+                pdf_file, save_all=True, append_images=images[1:], format="PDF"
+            )
+            return pdf_file.getvalue()
+
+    doc = documentai.Document
+    document = doc.from_json(json.dumps(gt_json))
+    synthesized_images = []
+    for i in range(len(document.pages)):
+        synthesized_images.append(decode_image(document.pages[i].image.content))
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:708]:
```suggestion
    document = documentai.Document.from_json(json.dumps(gt_json))
    synthesized_images = [decode_image(page.image.content) for page in doc.pages]
```

File: incubator-tools/best-practices/utilities/utilities.py [hunk 14]
```diff
@@ -0,0 +1,743 @@
...
+    # mime_type=temp_info['contentType']
+    document = {"content": image_content, "mime_type": "application/pdf"}
+
+    # Configure the process request
+    request = {"name": name, "raw_document": document, "skip_human_review": False}
+
+    # Recognizes text entities in the PDF document
+    result = client.process_document(request=request)
+
+    return result
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:743]:
```suggestion
def process_document_sample(
    project_id: str, location: str, processor_id: str, pdf_bytes: bytes, processor_version: str
):
    """This function is used to process the files using pdf bytes and provides the processed file"""
    # You must set the `api_endpoint` if you use a location other than "us".
    opts = ClientOptions(api_endpoint=f"{location}-documentai.googleapis.com")
    client = documentai.DocumentProcessorServiceClient(client_options=opts)

    name = client.processor_version_path(
        project_id, location, processor_id, processor_version_id
    )
    
    raw_document = documentai.RawDocument(content=pdf_bytes, mime_type="application/pdf")

    # Configure the process request
    request = documentai.ProcessRequest(
        name=name,
        raw_document=raw_document,
        skip_human_review=False
    )

    # Recognizes text entities in the PDF document
    result = client.process_document(request=request)

    return result
```

**Review by holtskinner — APPROVED** [review]
Some suggested updates for readability/efficiency.

File: incubator-tools/best-practices/utilities/utilities.py [hunk 15]
```diff
@@ -0,0 +1,743 @@
...
+    Returns:
+        documentai.Document: A DocumentAI Document proto representation of the downloaded JSON.
+    """
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(blob_name_with_prefix_path)
+
+    contents = blob.download_as_string()
+    doc = documentai.Document.from_json(contents.decode())
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:183]:
```suggestion
    doc = documentai.Document.from_json(blob.download_as_bytes())
```

File: incubator-tools/best-practices/utilities/utilities.py [hunk 16]
```diff
@@ -0,0 +1,743 @@
...
+
+    Returns:
+        dict: A dictionary representation of the downloaded JSON file.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blob = bucket.blob(blob_name)
+
+    contents = blob.download_as_string()
+    return json.loads(contents.decode())
```
Comment (holtskinner) [inline incubator-tools/best-practices/utilities/utilities.py:316]: Is there a need for this function and the `documentai_json_proto_downloader()` function? They seem to do the same thing

Comment (bharadwajreddyvangimalla) [reply incubator-tools/best-practices/utilities/utilities.py:315]: Blob downloader downloads the file as string while documentai_json_proto_downloader will download the json as document ai object

Comment (dmoonat) [reply incubator-tools/best-practices/utilities/utilities.py:315]: blob_downloader returns the json object and documentai_json_proto_downloader returns the document object

Comment (holtskinner) [reply incubator-tools/best-practices/utilities/utilities.py]: Ok, is there a reason that the notebook using `blob_downloader()` doesn't want a Document AI object? Can it use the document ai object? Or is this for reading in different types of JSON data?
